docs/contributing/coding-guidelines/ros-nodes/parameters.md

```diff
@@ -1,36 +1,150 @@
 # Parameters
 
-The ROS packages in Autoware have ROS parameters. You need to customize the parameters depending on your applications.
-It is recommended not to set default values when declaring ROS parameters to avoid unintended behaviors due to accidental use of default values.
-Instead, set parameters from configuration files named `*.param.yaml`.
+Autoware ROS nodes have declared parameters which values are provided during the node start up in the form of a parameter file. All the expected parameters with corresponding values should exist in the parameter file. Depending on the application, the parameter values might need to be modified.
 
-For understanding ROS 2 parameters, also check out the official documentation [Understanding parameters](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Parameters/Understanding-ROS2-Parameters.html).
+Find more information on parameters from the official ROS documentation:
 
-## Parameter files
+- [Understanding ROS 2 Parameters](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Parameters/Understanding-ROS2-Parameters.html)
+- [About ROS 2 Parameters](https://docs.ros.org/en/humble/Concepts/About-ROS-2-Parameters.html)
 
-Autoware has the following two types of parameter files for ROS packages:
+## JSON Schema
```

Comment:
As noted in [our previous discussion](https://github.com/orgs/autowarefoundation/discussions/3325#discussioncomment-5417178), we came to an agreement on locating the JSON Schema files in the `package/schema/default_node_name.schema.json` path.

However, it's important to acknowledge that Picknik's library doesn't incorporate this feature yet. And our CI actions are not yet fully prepared to accommodate the automatic association feature of VS Code through the `.vscode/settings.json` file. This is an aspect we all are aware of and should work towards.

Because of this, I'd propose that we consider introducing JSON Schema related documentation only after resolving these two hurdles. This approach would ensure we're not putting the cart before the horse, thus preventing potential confusion amongst our developers regarding the actual functionality of these JSON Schema files.

That being said, it could be beneficial to outline our prospective plans for the use of JSON Schema files in this documentation. This would provide a clear roadmap for the Autoware development team, giving us a sense of direction and a better understanding of the upcoming improvements to our workflow.

Reply:
@xmfcx @mitsudome-r @HamburgDave @oguzkaganozt 

The content of the contribution guidelines has already been approved, see https://github.com/orgs/autowarefoundation/discussions/3371#discussioncomment-5544013. 

Regarding the location of the JSON Schema, that's what the contribution guidelines state. From the commit:

> The schema file path is `INSERT_PATH_TO_PACKAGE/schema/` and the schema file name is `INSERT_NODE_NAME.schema.json`. To adapt the template to the ROS node, replace each `INSERT_...` and add all parameters `1..N`.

Or did I miss something?

Reply:
I appreciate the ongoing discussion and your continuous engagement in shaping our contribution guidelines, as per the shared agreement in [this thread](https://github.com/orgs/autowarefoundation/discussions/3371#discussioncomment-5544013).

The conversations have been extensive, and it's important to revisit and consolidate our shared understanding.

Our primary aim from the outset has been to establish a single source of truth, with a focus on omitting default values from the nodes. To that end, we've integrated ideas from both @kenji-miyake's [generate_parameter_library](https://github.com/PickNikRobotics/generate_parameter_library) and @doganulus' JSON Schema files, despite initial differing perspectives.

Our current consensus, as I understand, is to provide the flexibility to choose between using: 
- the 'generate_parameter_library'
- or standard ROS 2 parameters, provided the defaults are not set in the source files.

Furthermore, once the 'generate_parameter_library' can read from JSON Schema files, these files can be included in the schema folder.

[Your video from the DevOps Dojo: ROS Node Configuration - Final3 Proposal](https://user-images.githubusercontent.com/30662532/228092478-1a61b9d4-1109-40a5-973c-773eb53ac40b.mp4) illustrates the current requirement for manual input of the `.schema.json` file path.
![Screenshot from 2023-05-24 18-07-14](https://github.com/autowarefoundation/autoware-documentation/assets/10751153/ab784f5d-6abb-4c1d-87c5-c032187fc8b5)
- Here, you've made VS Code attempt to validate all YAML Param files (with `*.param.yaml`)
- If you try to open any other YAML file, it will be shown as invalidated
- For this to work correctly, user also needs to enter the full path of the `.yaml` file or files which this schema is trying to validate.

For enhanced user experience, I propose that we create a CI action to generate the `settings.json` file from the schema files. This way, users wouldn't need to manually specify these details, reducing the likelihood of typing errors. Once this is accomplished, the next step would be to validate the `.yaml` files using CI, based on the schema files.

Only after we've taken these steps should we consider adding this behavior to the documentation and recommending developers to adopt it. Doing so prematurely may lead to confusion without substantial benefit.

Reply:
There was a lot of discussion whether or not use PickNikRobotics library and in the end, it did not make the final proposal. However, the final proposal doesn't close the door on using PickNikRobotics library further down the road.

We decided that even though we don't utilize the JSON Schema immediately, creating it when doing the other work on the ROS nodes made sense and was decided on (without any objections) in the Open AD Kit meeting on April 6th. I pinged other AWF channels and individuals weeks ahead to get input.

Without the JSON Schema file you cannot do any validation regardless. It's a chicken and egg problem. An CI action can be created at any time and the work is outlined here:
- https://github.com/autowarefoundation/autoware-github-actions/issues/232

Create an auto-generated `settings.json` file:
- https://github.com/autowarefoundation/autoware/issues/3424

The JSON Schema is used in VSCode to help a developer understand what parameters are required for a node. To use VSCode is not obligatory and therefore it is under "Tips and Tricks", i.e., not part of the requirement, but as a tool to help our developers.

Can we please merge the new coding guidelines which have already been approved?

Reply:
I appreciate your commitment to refining our coding guidelines, and the rationale you've provided makes sense. However, I would like to express some concerns that arise from our collective discussions.

While we had multiple threads 
- [ROS Node Configuration](https://github.com/orgs/autowarefoundation/discussions/3260),
- [ROS Node Configuration - Final Proposal](https://github.com/orgs/autowarefoundation/discussions/3288),
- [ROS Node Configuration - Final2 Proposal](https://github.com/orgs/autowarefoundation/discussions/3325), 
- [DevOps Dojo: ROS Node Configuration - Final3 Proposal](https://github.com/orgs/autowarefoundation/discussions/3371),

my last comment in the Final2 Proposal [here](https://github.com/orgs/autowarefoundation/discussions/3325#discussioncomment-5417178) led to a unified conclusion from the remaining participants, but a new thread was then started which saw diminished engagement. This raises questions about the representation of viewpoints in the final proposal.

In the Open AD Kit meeting on April 6th, the decision to create JSON Schema files, despite their immediate non-utilization, was passed without objections. I acknowledge the effort you made to gather input weeks ahead.

The process this proposal underwent isn't as important to me as its content and potential impact on the Autoware repository. An uncontested approval should not deter further discussion on an idea, particularly one of this magnitude.

Regarding the 'chicken and egg' analogy, my perspective differs. I view this situation as more akin to 'putting the cart before the horse'. The initial schema file ought to be created concurrently with a supporting CI action. Currently, we lack a mechanism that validates the schema files themselves, and manually entering full paths for each file on VS Code is cumbersome.

In the documentation, we can indeed outline future plans about this feature, but as it stands, the existing provisions are insufficient for productive use. A more practical approach is needed before we can integrate these guidelines into our workflow.

We can discuss this in the next Open AD Kit meeting or raise it to the next TSC meeting if necessary.

Reply:
@xmfcx I think a supporting CI action for schema automation is nice but I suggest not stating it as a blocking item. It is not that hard to enter full paths for each node by hand. Considering node names are not changing frequently, this is probably a one-time effort for each node. So even it may not need such automation initially. Automation may be needed to publish Autoware schemas at a remote endpoint as many cloud-native software do. 

One thing I want from the software group or the TSC is to standardize/canonize the paths of nodes across Autoware or let me know if such a decision is already in place. If you can give a guarantee to have paths like below and ensure it across Autoware,

```
`<prefix>`/`<component>`/`<package>`/{`include`, `src`, `config`, `schema`, `tests`}
```

Then most Autoware developers would not need to touch `vscode` settings once again. We need some conventions to keep it ordered and systematic.

Reply:
@doganulus thank you for your response. While I agree that it's possible for us to manually input full paths, it's important to remember that each package can potentially have multiple nodes, and each node can carry multiple parameter files.
**Examples:**
 - https://github.com/autowarefoundation/autoware.universe/blob/main/planning/behavior_path_planner/config/dynamic_avoidance/dynamic_avoidance.param.yaml
 - https://github.com/autowarefoundation/autoware.universe/blob/main/planning/behavior_path_planner/config/lane_change/lane_change.param.yaml

Considering this, the manual process can become significantly cumbersome and raise the maintenance cost.

It's true that node names don't frequently change, yet we should be mindful that changes can and do occur. Keeping the `settings.json` file updated in light of these changes would necessitate an ongoing maintenance effort, potentially diverting our attention from core development tasks.

In terms of standardizing the paths, I agree that conventions are crucial for maintaining an orderly and systematic workflow. We can potentially standardize paths for main components such as "planning, sensing, perception, etc.". However, given that nodes and packages are regularly updated and modified, the primary focus should be on developing a CI action that can create the `settings.json` file automatically.

With the absence of a validation process for `.schema.json` files currently, I strongly believe that having the automation in place before we advance is key to maintaining a manageable workload and a smooth workflow. Hence, I see this as a blocking item. And that automation should precede further advancement.

Reply:
@xmfcx What do you expect from the `.schema.json` validation? Do you want to check the `schema.json` files to satisfy, say, the `draft-07` specification?

Reply:
> What do you expect from the .schema.json validation? Do you want to check the schema.json files to satisfy, say, the draft-07 specification?

@doganulus, you are correct. At its core, I would indeed want the `.schema.json` files to conform to specifications such as the `draft-07`, which is an established and reliable standard for JSON Schema.

Currently, we can manually check these files using tools like VS Code or potentially some online validators. However, this manual process isn't scalable or efficient in the long run.

Moreover, I envision a system that not only validates the `.schema.json` files but also employs these schemas to automatically validate the `.yaml` files. Such a system would greatly enhance our efficiency and reliability, ultimately reducing the likelihood of errors slipping through manual checks. Automation here serves a dual purpose - standardization and quality assurance.

Reply:
Hi, I just noticed this now, has it changed from the conclusion before? My understanding is:
- Creating a parameter definition file is recommended (but not required).
- When creating a parameter definition file, follow either the JSON schema or PickNik's generate_parameter_library format.

Reply:
> Hi, I just noticed this now, has it changed from the conclusion before? My understanding is:
> - Creating a parameter definition file is recommended (but not required).

@isamu-takagi this is correct.

> - When creating a parameter definition file, follow either the `JSON schema` or PickNik's `generate_parameter_library` format.

As for your second point, it's partially accurate. While we recommend following either the JSON Schema or PickNik's `generate_parameter_library` format, we must consider the current limitations. Implementing `.schema.json` files right now, without the necessary supporting mechanisms, could lead to additional maintenance efforts with limited advantages.

Therefore, I propose that we proceed with adding .schema.json files only when:

1. CI actions are in place to:
  - Validate the schema files (essential)
  - Validate the .yaml files with the schemas (essential)
  - Create settings.json files to support VS Code (desirable but not critical)
2. The `generate_parameter_library` can accommodate JSON Schemas, ensuring we do not end up with multiple sources of truth (desirable but not critical)

Without these capabilities, we risk having superfluous schema files in the repository, functional only when manually added to the VS Code's `settings.json` file.

In light of these considerations, I suggest that this proposal be implemented concurrently with the above-mentioned functionalities to maximize its effectiveness and efficiency.

Reply:
@xmfcx Sorry, I didn't understand about the validation and settings.json of VS Code. I think we are trying to do something more complicated than originally thought. For the time being, parameter definition file is a kind of README. Developers write parameter definition file instead of writing the parameters section of README.

So there is no particular difference between formats, maintenance costs are paid by the person who chooses the format. The only thing that matters to anyone other than the developer of that package is reading the parameter definition file. Of course, definition files are more convenient for those who want their tools to process the parameters.

Reply:
Or does someone other than the package maintainer add parameter definition files?

Reply:
@xmfcx 
I'm with @isamu-takagi -san on this.

The original reason that Kasper wanted to include JSON schema was to have a single source of truth about parameter definition. My understanding is that we have come to agreement in the GitHub Discussion to use JSON schema on that, and I think having that consensus defined in coding-guideline could happen regardless of the the existence of CI. Also, regarding the parameter generation using PickNick was decided to be optional thing that is up to the user whether they want to use it or not.

Is your concern that you don't want to have people creating JSON schema file that might not be valid by creating this PR public? In that case, we might be a valid concern and we can consider CI with the GitHub Actions, but I'm not sure why you want to have parameter generation feature available before having this guideline merged.

Reply:
@mitsudome-r said:
> Is your concern that you don't want to have people creating JSON schema file that might not be valid by creating this PR public?

That is one concern. And the other is, that it wouldn't have a use. That's why I've proposed adding it when it actually has a use.

> but I'm not sure why you want to have parameter generation feature available before having this guideline merged.

In [my previous post above](https://github.com/autowarefoundation/autoware-documentation/pull/367/files#r1205435380), I've stated

1. CI actions are in place to:
  - Validate the schema files (essential)
  - Validate the `.yaml` files with the schemas (essential)
  - Create `settings.json` files to support VS Code (desirable but not critical)
2. The `generate_parameter_library` can accommodate JSON Schemas, ensuring we do not end up with multiple sources of truth (desirable but not critical)

So, to me, parameter generation is **not** a necessity to add this.

@isamu-takagi said:
> For the time being, parameter definition file is a kind of README. Package maintainers write parameter definition file instead of writing the parameters section of README.

I looked into it to be more than just a Readme file. That's why I've raised my points.

If we are going to see it that way, I am also OK with this. Then I can just treat it like any other text file.

If there are no more objections, I can mark this conversation as resolved.

Reply:
>We have a list of the nodes which need to be updated, find the spreadsheet [here](https://docs.google.com/spreadsheets/d/1yAnXedY3zv2oD1prDZ_AyYiFMI8T9ZVrRNqm9kEPYcs/edit?usp=sharing) and look at the tab "ROS Node Config". @mitsudome-r has gone through the list to remove any ROS package which doesn't need to be modified.
>
>Edit: so to answer your question, I think anyone can make the changes but the package maintainer should probably review the changes to ensure it looks OK.

I understood @xmfcx -san's concern. So there are cases someone will introduce a JSON schema to the packages I maintain. It goes against my understanding. It is forced to create a definition file and its format is fixed.

To be acceptable, leave schema file maintenance to the schema author rather than the package maintainer. In other words, I don't update the schema file when I update my package's parameters. The author of the schema file will update it if necessary and I will review it. Then it's OK for me.

Reply:
@isamu-takagi It is better that the node maintainer maintains the schema. No one else knows node parameters better than the node maintainer. I don't know the minimum and maximum values for your parameters or any other detail. 

JSON schema is written in the standard JSON and is widely used as an Internet standard with many benefits now or in the future. We have discussed this in great detail. Please see the previous discussions. This cannot be simpler than that. 

What @xmfcx mentions is about how to reap these benefits further. This doesn't bring an additional load to the maintainer. Just write the schema and enjoy benefits...